File: Retos/Reto #6 - PIEDRA, PAPEL, TIJERA, LAGARTO, SPOCK [Media]/dart/malopezrom.dart
/*
 * Crea un programa que calcule quien gana mΓ‘s partidas al piedra,
 * papel, tijera, lagarto, spock.
 * - El resultado puede ser: "Player 1", "Player 2", "Tie" (empate)
 * - La funciΓ³n recibe un listado que contiene pares, representando cada jugada.
 * - El par puede contener combinaciones de "πΏ" (piedra), "π" (papel),
 *   "βοΈ" (tijera), "π¦" (lagarto) o "π" (spock).
 * - Ejemplo. Entrada: [("πΏ","βοΈ"), ("βοΈ","πΏ"), ("π","βοΈ")]. Resultado: "Player 2".
 * - Debes buscar informaciΓ³n sobre cΓ³mo se juega con estas 5 posibilidades.
 */


/**
 * Funcion main con los casos de ejemplo.
 */
main() {

  var game = Game([["πΏ", "βοΈ"],
  ["βοΈ", "πΏ"],
  ["π¦", "βοΈ"],
  ["πΏ", "π"],
  ["π", "πΏ"],
  ["πΏ", "πΏ"],
  ["π", "π¦"],
  ["π", "βοΈ"]],"Player 1","Player 2");
  print("El resultado es : ${evaluateGame(game)}");

  var game2 = Game([["πΏ", "βοΈ"],
      ["π¦", "βοΈ"],
      ["π", "πΏ"],
      ["π", "π"],

      ],"Player 1","Player 2");
  print("El resultado es : ${evaluateGame(game2)}");


  var game3 = Game(
      [
        ["π", "π¦"],
        ["π¦", "π"],
        ["π", "π"],

      ], "Player 1", "Player 2");
  print("El resultado es : ${evaluateGame(game3)}");

}



/*
Objeto que representa las opciones de juego.
*/
var options = ["πΏ","π","βοΈ","π¦","π"];
/**
 * Enumerado que representa un resultado de una jugada.
 */
enum Result {Tie,Player1, Player2}

/**
 * Combinaciones de jugadas.
 */
Map<String,Map<String,Result>> combinations =
{ //Piedra
  "πΏ": {"πΏ": Result.Tie, "π": Result.Player2, "βοΈ": Result.Player1, "π¦": Result.Player1, "π": Result.Player2},
  //Papel
  "π": {"πΏ": Result.Player1, "π": Result.Tie, "βοΈ": Result.Player2, "π¦": Result.Player2, "π": Result.Player1},
  //Tijera
  "βοΈ": {"πΏ": Result.Player2, "π": Result.Player1, "βοΈ": Result.Tie, "π¦": Result.Player1, "π": Result.Player2},
  //Lagarto
  "π¦": {"πΏ": Result.Player2, "π": Result.Player1, "βοΈ": Result.Player2, "π¦": Result.Tie, "π": Result.Player1},
  //Spock
  "π": {"πΏ": Result.Player1, "π": Result.Player2, "βοΈ": Result.Player1, "π¦": Result.Player2, "π": Result.Tie}
};


/**
 * Clase que representa una jugada
 */
class Game{
  String player1;
  String player2;
  List<List<String>> game;
  Game(this.game,this.player1, this.player2);

}

/**
 * Funcion que busca el ganador si lo hay de una lista de jugadas
 * @param listOfGame Listado de jugadas
 * @return Resultado de la jugada
 * @see Result
 */
Result foundWinner(List<Result> game){
  var player1= game.where((element) => element==Result.Player1).length;
  var player2= game.where((element) => element==Result.Player2).length;

  if(player1>player2){
    return Result.Player1;
  }
  else if(player2>player1){
    return Result.Player2;
  }
  else{
    return Result.Tie;
  }
}


/**
 * Funcion que evalua una partida de piedra, papel, tijera , lagarto o spock
 * @param game objeto que representa un juego con los jugadores y las jugadas de cada uno
 * @return String con el nombre del jugador o empate en el que no hay ningun ganador
 */

String evaluateGame(Game game) {
  List<Result> results = [];
  for (var e in game.game) {
    if (combinations.containsKey(e[0]) && combinations.containsKey(e[1])) {
      if (combinations[e[0]]![e[1]] == Result.Player1) {
        results.add(Result.Player1);
      } else if (combinations[e[0]]![e[1]] == Result.Player2) {
        results.add(Result.Player2);
      }
    }
  }

  switch(foundWinner(results)){
    case Result.Player1:
      return game.player1;
    case Result.Player2:
      return game.player2;
    case Result.Tie:
      return "empate";
  }




}

File: Retos/Reto #6 - PIEDRA, PAPEL, TIJERA, LAGARTO, SPOCK [Media]/python/malopezrom.py
# /*
# * Crea un programa que calcule quien gana mΓ‘s partidas al piedra,
# * papel, tijera, lagarto, spock.
# * - El resultado puede ser: "Player 1", "Player 2", "Tie" (empate)
# * - La funciΓ³n recibe un listado que contiene pares, representando cada jugada.
# * - El par puede contener combinaciones de "πΏ" (piedra), "π" (papel),
# *   "βοΈ" (tijera), "π¦" (lagarto) o "π" (spock).
# * - Ejemplo. Entrada: [("πΏ","βοΈ"), ("βοΈ","πΏ"), ("π","βοΈ")]. Resultado: "Player 2".
# * - Debes buscar informaciΓ³n sobre cΓ³mo se juega con estas 5 posibilidades.
# */

# Opciones de juego
options = {"πΏ", "π", "βοΈ", "π¦", "π"}

# Resultado de la jugada
# 0 = empate
# 1 = gana player 1
# 2 = gana player 2
resultado = {0, 1, 2}

# Matriz de combinaciones posibles
combinaciones = { # piedra
                "πΏ": {"πΏ": 0, "π": 2, "βοΈ": 1, "π¦": 1, "π": 2},
                # papel
                "π": {"πΏ": 1, "π": 0, "βοΈ": 2, "π¦": 2, "π": 1},
                # tijera
                "βοΈ": {"πΏ": 2, "π": 1, "βοΈ": 0, "π¦": 1, "π": 2},
                # lagarto
                "π¦": {"πΏ": 2, "π": 1, "βοΈ": 2, "π¦": 0, "π": 1},
                # spock
                "π": {"πΏ": 1, "π": 2, "βοΈ": 1, "π¦": 2, "π": 0}
                }


# Clase que representa un juego
# list = lista de jugadas
# player1 = nombre del jugador 1
# player2 = nombre del jugador 2
class Game:
    def __init__(self, list, player1, player2):
        self.list = list
        self.player1 =player1
        self.player2 = player2


# /**
# * Funcion que busca el ganador si lo hay de una lista de jugadas
# * @ param listOfGame Listado de jugadas
# * @ return Resultado de la jugada
# * @ see Result
# */
def found_winner(game:Game):
    player1 = game.list.count(lambda: resultado[1])
    player2 = game.list.count(lambda: resultado[2])
    if player1 > player2:
        return game.player1
    elif player1 < player2:
        return game.player2
    else:
        return 0


# /**
# * Funcion que evalua una partida de piedra, papel, tijera
# * @ param game objeto que representa un juego con los jugadores y las jugadas de cada uno
# * @ return String con el nombre del jugador o empate en el que no hay ningun ganador
# */
def evaluate_game(game: Game) -> resultado:
    player1 = 0
    player2 = 0
    for i in game.list:
        if i[0] in options and i[1] in options:

            if combinaciones[i[0]][i[1]] == 1:
                player1 += 1
            elif combinaciones[i[0]][i[1]] == 2:
                player2 += 1
            else:
                player1 += 0
                player2 += 0

    if player1 > player2:
        return game.player1
    elif player1 < player2:
        return game.player2
    else:
        return "Empate"




#################### casos de prueba ####################

game1 = Game([["πΏ", "βοΈ"],
             ["βοΈ", "πΏ"],
             ["π¦", "βοΈ"],
             ["πΏ", "π"],
             ["π", "πΏ"],
             ["πΏ", "πΏ"],
             ["π", "π¦"],
             ["π", "βοΈ"]], "Player 1", "Player 2")
print("El resultado es : " + evaluate_game(game1))

game1 = Game([["πΏ", "βοΈ"],
              ["π¦", "βοΈ"],
              ["π", "πΏ"],
              ["π", "π"]], "Player 1", "Player 2")
print("El resultado es : " + evaluate_game(game1))

game1 = Game([["π", "π¦"],
              ["π¦", "π"],
              ["π", "π"]], "Player 1", "Player 2")
print("El resultado es : " + evaluate_game(game1))
